### PR TITLE
[PD-2046] misc protections

### PR DESCRIPTION
**Pull request by lemniscate613**
This PR prevents users from:
* Editing Admin role
* Removing Admin role from the last person with this role in a company

I also removed ~75 PEP8 errors while waiting for Jenkins.

# Review

If using Docker:

```
dkm test myjobs.tests.test_manage_users
```

Then visit: https://secure.my.jobs/manage-users/#/roles

There is not "Edit" button for the Admin role:

screen shot 2016-01-11 at 1 43 39 pm

And if you try to review the Admin role from the only user with that role, you get an error in the UI:

screen shot 2016-01-11 at 1 43 33 pm

### CODE REVIEW

File: myjobs/views.py [hunk 1]
```diff
@@ -1208,12 +1211,34 @@ def api_edit_user(request, user_id=0):
...
+        # 1. Is Admin not in assigned_roles?
+        if "Admin" not in assigned_roles:
+            # 2. Does this company only have one user assigned to Admin role?
+            admin_role = (Role
+                          .objects
+                          .filter(company=company)
+                          .filter(name="Admin"))
+            users_with_admin = User.objects.filter(roles__id=admin_role[0].id)
+
+            if len(users_with_admin) <= 1:
```
Comment (tdphillips) [inline myjobs/views.py]: @JLMcLaugh had some concerns about this during our team meeting. Were they discussed?

Comment (lemniscate613) [inline myjobs/views.py]: Jason M and I spoke with @fezick about this after that meeting. He said this is correct -- require that each company have at least once user be assigned the Admin role.

File: myjobs/views.py [hunk 2]
```diff
@@ -1208,12 +1211,34 @@ def api_edit_user(request, user_id=0):
...
+                # 3. Is that one user this user?
+                if unicode(users_with_admin[0].id) == user_id:
+                    ctx["success"] = "false"
+                    ctx["message"] = "To remove the Admin role from this user \
```
Comment (tdphillips) [inline myjobs/views.py]: Continuing like this produces the following string.

```
'To remove the Admin role from this user
                                       you must first assign it to another user.
                                      Every company must have at least one
                                       user with the Admin role.'
```

You might want something more like

```
                               ctx["message"] = ("To remove the Admin role from this user, "
                                                 "you must first assign it to another user. "
                                                 "Every company must have at least one "
                                                 "user with the Admin role.")

Comment (lemniscate613) [inline myjobs/views.py]: I didn't know you could do that. Thank you. :+1:

File: myjobs/views.py [hunk 3]
```diff
@@ -1208,12 +1211,34 @@ def api_edit_user(request, user_id=0):
...
+                          .filter(company=company)
+                          .filter(name="Admin"))
+            users_with_admin = User.objects.filter(roles__id=admin_role[0].id)
+
+            if len(users_with_admin) <= 1:
+                # 3. Is that one user this user?
+                if unicode(users_with_admin[0].id) == user_id:
+                    ctx["success"] = "false"
+                    ctx["message"] = ("To remove the Admin role from this user "
+                                      "you must first assign it to another
```
Comment (tdphillips) [inline myjobs/views.py]: Missing a quote on the end.

File: seo/tests/test_views.py [hunk 4]
```diff
@@ -1095,11 +1100,10 @@ def test_custom_moc(self):
...
 
-        default_job.update({
-                    'buid':'1',
-                    'mapped_moc':None,
-                    'mapped_mocid':None,
-                    'onet':'1234'})
+        default_job.update({'buid': '1',
+                            'mapped_moc': None,
+                            'mapped_mocid': None,
+                            'onet': '1234'})
```
Comment (koonsies) [inline seo/tests/test_views.py]: I'm not sure what is going on with these test changes. Is it mentioned in the summary?

Comment (tdphillips) [inline seo/tests/test_views.py]: Looks like PEP 8.

Comment (koonsies) [inline seo/tests/test_views.py]: Yup. Definitely a Monday.

**Comment by koonsies** [comment]
:+1: Anyone have objections to merging this?

File: myjobs/views.py [hunk 5]
```diff
@@ -1129,7 +1131,7 @@ def api_create_user(request):
             roles = request.POST.getlist("assigned_roles[]", "")
             # Create list of role_ids from names
             for i, role in enumerate(roles):
```
Comment (aspidites) [inline myjobs/views.py]: This for loop can be replaced by:
```python
role_ids = Role.objects.filter(name__in=roles, company=company).values_list("pk", flat=True)
```

File: myjobs/views.py [hunk 6]
```diff
@@ -1208,12 +1211,35 @@ def api_edit_user(request, user_id=0):
             return HttpResponse(json.dumps(ctx),
                                 content_type="application/json")
 
+        # Companies must have at least one user assigned to an Admin role.
+        # Check that -- if this user is edited -- this will be the case.
+        # 1. Is Admin not in assigned_roles?
```
Comment (aspidites) [inline myjobs/views.py]: Everything from `if "Admin" not in assigned_roles:"` to `if len(users_with_admin)` can be replaced by:

```python
user_with_admin = User.objects.filter(roles__name="Admin", roles__company=company)
```

**Comment by misterjonscott** [comment]
Looks good.

**Comment by tdphillips** [comment]
The outstanding comments have performance implications that I would like to see resolved at some point. Doing a number of database queries equal to len(roles) is len(roles) times slower than doing one filter.

**Pull request merged**